Fixed Position encoder for speed up

IRiViI opened the pull request:
Hey,

First of all, happy new year and stuff.

I made some small changes to the code such that the code will be 100x faster when you are dealing with a fixed input size.
How it works:
- It generates an positional encoding once
- It repeats it for the batch size (and changes it if needed)

The class is called FixedPositionalEncoding2D. I made an additional class SinglePositionalEncoder2D because the code is very similar to the original version. And how they both reference to that class to reduce duplicate code.
If you want to test the speed difference you can call the "test_torch_2D_fixed_speedup" function in the test suite.

Let me know if you like it. I didn't create a readme part yet because maybe you hate the idea hehe ;).

Cheers!

IRiViI commented:
I probably butchered the tests due to putting the speed test function over there. I can remove it if you think this would be great addition.

tatp22 commented on positional_encodings/positional_encodings.py:
```diff
@@ -76,7 +76,40 @@ def forward(self, tensor):
+        self.repeated_pos_encoding = self.pos_encoding.repeat(32, 1, 1, 1)
+
+    def forward(self, tensor):
+        # Ensure encoding on same device
+        if self.device is not tensor.device:
+            self.pos_encoding = self.pos_encoding.to(tensor.device)
+            self.device = self.pos_encoding.device
```
Would just doing `self.pos_encoding.to(tensor.device)` not work? from the docs:
https://pytorch.org/docs/stable/generated/torch.Tensor.to.html#torch.Tensor.to

```
If the self Tensor already has the correct torch.dtype and torch.device, then self is returned. Otherwise, the returned tensor is a copy of self with the desired torch.dtype and torch.device.
```

IRiViI replied:
You are probably right about this! I don't know which of the two is the better one. I will test if it makes an impact on the speed, if not, I think your solution does indeed look cleaner. 

The problem is that the pos_encoding is not automatically on the same device as the model. I which there is some method which makes this tensor move when you perform: model.to(device). If this is possible I think this would be the best solution. Do you know if this is possible?

tatp22 replied:
I am not sure there is something like that :cry:. That would indeed be the best...

IRiViI replied:
Yeah, it feels kind of dirty. Also, it could result into and error atm if someone would switch from device of the data. This because I left out the device check in the newer version because it affect the speed quite a bit if I remember correctly. Maybe I could add the if statement again because that one was a bit less taxing if I think. But not today.

tatp22 commented on positional_encodings/positional_encodings.py:
```diff
@@ -76,7 +76,40 @@ def forward(self, tensor):
+        self.pos_encoder = SinglePositionalEncoding2D(channels)
+        self.pos_encoding = self.pos_encoder(torch.ones(1, shape[0], shape[1], channels))
+        self.device = self.pos_encoding.device
+        self.repeated_pos_encoding = self.pos_encoding.repeat(32, 1, 1, 1)
```
is 32 just arbitrary?

tatp22 replied:
We could make it an input parameter probably...

IRiViI replied:
32 is indeed out of thin air. I needed an initial batch size and 32 is a very standard batch size so I went for that one. It really doesn't matter at all what you do here because it will be overwritten the first time you call the model. But yeah I agree that adding a key parameter would look nicer.

tatp22 replied:
Anyways I think what you did is good :+1:

tatp22 commented on test_suite.py:
```diff
@@ -103,6 +103,23 @@ def test_torch_summer():
+    start_time = time.time()
+    for i in range(sample_size):
+        pos_enc(data)
+    print(f"Duration of original: {time.time()- start_time}")
```
I am not so sure that the print does anything when ran with `pytest -vv`, but i could be wrong.

In any case, it would be cool to have an assert here on something. I am not sure on what exactly, since i don't think asserting on time is a good idea (since the test will probably be flaky), but still it would be cool to have something here.

IRiViI replied:
yeah, this was just a place holder in order to let you test the speed up. But I have the idea to confirm that the output of the fixed version matches the output of the original version. What do you think of that test?

tatp22 replied:
I see you did that already, nice job :+1:

IRiViI replied:
Thank you

tatp22 commented on positional_encodings/positional_encodings.py:
```diff
@@ -76,7 +76,40 @@ def forward(self, tensor):
+            self.repeated_pos_encoding = self.pos_encoding.repeat(tensor.shape[0], 1, 1, 1)
+        return self.repeated_pos_encoding
+
+class PositionalEncoding2D(nn.Module):
```
do we need this class here? all it is doing is just setting up the positional encoding class from before and then just calling its forward method in this forward method.

My suggestion is to get rid of this class, and to leave the name of the other one, while just having the new one called `Fixed`...

IRiViI replied:
I think it would even be cleaner to merge the FixedPositonalBlaBla with the original PositinalBlaBla and have a key argument of shape. And if you provide shape, you will be using the fixed route and if you leave the shape you will be using the original route. So just 1 class instead of 3.

tatp22 commented on positional_encodings/positional_encodings.py:
```diff
@@ -76,7 +76,40 @@ def forward(self, tensor):
-        return emb[None, :, :, :orig_ch].repeat(batch_size, 1, 1, 1)
+        return emb[None, :, :, :orig_ch]
+
+class FixedPositionalEncoding2D(nn.Module):
+    def __init__(self, shape, channels):
+        super(FixedPositionalEncoding2D, self).__init__()
+        self.pos_encoder = SinglePositionalEncoding2D(channels)
+        self.pos_encoding = self.pos_encoder(torch.ones(1, shape[0], shape[1], channels))
```
This could probably work for all positional encodings! I could imagine the function call looking something like this:

```python3
# pos_encoder = PEnc2D(channels)
def __init__(self, pos_encoder, shape): #shape would include channels
  super(...)
  self.pos_encoder = pos_encoder
  self.pos_encoding = self.pos_encoder(torch.ones(1, **shape))
```

IRiViI replied:
Mmm this is indeed maybe even nicer. So have a wrapper class of sorts which you can use for all 3 methods.  Sounds good.

tatp22 replied:
I saw you added that, nice :smile_cat:

IRiViI replied:
Yep, is a good idea of your to do it this way.

tatp22 reviewed (commented):
Hey @IRiViI, thank you for your PR!

So from what i understand correctly, what this PR does is that instead of making a positional encoding on every forward pass, we create one with the class initialization, and then based on the batch size, we repeat this penc for as many entriies as there are in the batch? (if it is different that is)

Very cool! I think that this could save some compute power for people that know how their data looks beforehand.

I think that your method could be made to fit with any of the pencs, actually, and it doesn't have to just be 2D. i wrote some comments and i want to know what you think about it.

Also, the reason that the CI doesn't pass is not because pytest fails, but because the code has to be styled correctly :wink: You can run `isort .; black .` from the top directory and that should do the trick.

IRiViI commented:
Good suggestions! Glad you like the idea. And you are right how it works. It just makes the positional encoding vector fixed. I was kind surprised that the ".repeat(batch_size,1,1,1)" is the most computational intensive part. So when I only fix the positional encoding code but still leave the .repeat part, it resulted in a 2x speed up. And after I also fixed the batch size (by checking if the batch size is the same) it made the code 50x faster again (so 100x).

IRiViI commented:
- Turned the class into a wrapper class. 
- Added 3 tests for the 1D, 2D and 3D encoders. 
- I didn't made it compatible with the permuted once though... 
- I also added a parameter "org_channels" to the three classes in order to obtain this value easily.
- O and the code is now even faster! yaaay. Now it's like a 200x or 600x speed up instead of the measly 100x. 

Fun fact: 
Adding "if self.something == None" Made my code 40x slower! Sooo I guess I should never use None in my tensor code

To do:
- I can make it compatible with the permuted once
- Add tekst in the documentation

IRiViI commented:
Oke, I added documentation in the Readme. I kept it short.
I also tried to work with the permuted classes but I just kept making mistakes so maybe you could add it?.. ;).
I added a getter to those permuted classes because I thought it would be the nicest way. I can remove that if you do not agree.

IRiViI commented:
O and let me know if you do  not like the name. I paid little attention to naming the class. FixEncoding might not be the nicest

tatp22 commented:
It's really nice these speedup numbers! I'll take a look at the name too and let you know if I come up with anything better.

I can take a look at the permute classes, but idk if i will get to them anytime soon (i can make an issue).

tatp22 commented:
I added issues for the remaining things. And i'll reformat the code after i merge it :wink:

IRiViI commented:
Oke cool!
